### ISSUE TITLE
Improve Auto-Analyze Mechanism for Better Concurrency

### ISSUE THREAD

**Issue by Rustin170506** [pingcap/tidb#55618]
## Background
The current implementation of `AnalyzeHighestPriorityTables` uses a `WaitGroupWrapper` to execute analysis jobs concurrently. This approach lacks flexibility in managing running jobs.

## Proposed Changes
1. Implement a dynamic job management system that maintains a constant number of running jobs up to the maximum concurrency limit.
2. Add functionality to track currently running analysis jobs by table ID.
3. Allow for dynamic adjustment of concurrency based on the `AutoAnalyzeConcurrency` parameter.
4. Replace the `WaitGroupWrapper` with individual goroutines for each job.

## Implementation Details
1. Add a `runningJobs` map to the `Refresher` struct to track active jobs.
2. Modify `AnalyzeHighestPriorityTables` to:
   - Check the current number of running jobs before starting new ones.
   - ~~Use goroutines for each job instead of a `WaitGroupWrapper`.~~
   - Dynamically start new jobs as others complete, maintaining the desired concurrency.
3. Implement proper locking mechanisms to ensure thread-safety when managing the job queue and running jobs.
4. Add a method to get the current running jobs.

## Expected Benefits
- More efficient use of system resources by maintaining a constant number of running jobs.
- Improved flexibility in managing analysis jobs, allowing for better responsiveness to system changes.
- Enhanced ability to monitor and control the auto-analyze process.
- Prepare for the new priority queue.

**Comment by hawkingrei** [comment]
What does ``` Use goroutines for each job```? The goroutine pool will be used to run the analysis. it is dangerouse. 

Historically, TiDB has experienced instances where a panic occurred on a goroutine, was recovered, and then panicked again, ultimately causing the entire TiDB to crash.

**Comment by hawkingrei** [comment]
Increasing the complexity of this is unnecessary.

**Comment by hawkingrei** [comment]
Also, this is my feature. If you have any questions or improvement，please contact me first.

**Comment by Rustin170506** [comment]
> What does ` Use goroutines for each job`? The goroutine pool will be used to run the analysis. it is dangerouse.

That was a wrong statement. I thought I would use a goroutine directly, but actually, I just used the wait group. You can take a look at the PR.

**Comment by Rustin170506** [comment]
> Also, this is my feature.

🥹 I just tried to implement the thing I need for the new priority queue. I almost need to rewrite the whole refresher. So I choose to add by myself. You even have closed the issue for that feature. So I decided to open a new issue to add it.

**Comment by hawkingrei** [comment]
> > What does ` Use goroutines for each job`? The goroutine pool will be used to run the analysis. it is dangerouse.
> 
> That was a wrong statement. I thought I would use a goroutine directly, but actually, I just used the wait group. You can take a look at the PR.

You made something simple sound too complicated.

**Comment by Rustin170506** [comment]
> You made something simple sound too complicated.

😅
Your token limiter（84 lines）:
<img width="648" alt="image" src="https://github.com/user-attachments/assets/9a7a9764-24f3-4c19-8b25-ffaeeb69173e">

My pr(Dynamic + tracking all running tables, 139 - 13 = 126 lines):
<img width="781" alt="image" src="https://github.com/user-attachments/assets/17ad8dfc-4859-4972-9b6d-010ba683c031">

**Comment by Rustin170506** [comment]
> Historically, TiDB has experienced instances where a panic occurred on a goroutine, was recovered, and then panicked again, ultimately causing the entire TiDB to crash.

https://github.com/pingcap/tidb/blob/d041cd054cfd8b7438b3031976b9b4fbd3dfba4e/pkg/statistics/handle/autoanalyze/refresher/refresher.go#L106

You used the normal wait group. 

```go
// Run runs a function in a goroutine, adds 1 to WaitGroup
// and calls done when function returns. Please DO NOT use panic
// in the cb function.
func (w *WaitGroupWrapper) Run(exec func()) {
	w.Add(1)
	go func() {
		defer w.Done()
		exec()
	}()
}
```

😅
